File: clients/go/node/store/reorg.go
package store

import (
	"fmt"
	"math/big"

	"rubin.dev/node/consensus"
	"rubin.dev/node/crypto"

	bolt "go.etcd.io/bbolt"
)

// ReorgToTip performs the Phase 1 disconnect/connect procedure to move the applied tip
// from the manifest tip to newTipHash (which must be present in block_index_by_hash).
//
// This mutates persistent chainstate (utxo + undo + manifest). It is deterministic given
// the stored blocks and index.
func (d *DB) ReorgToTip(p crypto.CryptoProvider, chainID [32]byte, newTipHash [32]byte, opts ApplyOptions) error {
	if d == nil || d.db == nil || d.manifest == nil {
		return fmt.Errorf("db not ready")
	}
	if p == nil {
		return fmt.Errorf("crypto provider required")
	}

	oldTipHash, err := parseHex32(d.manifest.TipHashHex)
	if err != nil {
		return err
	}
	if oldTipHash == newTipHash {
		return nil
	}

	forkHash, err := d.findForkPoint(oldTipHash, newTipHash)
	if err != nil {
		return err
	}

	// Disconnect old chain tip -> fork+1.
	cur := oldTipHash
	for cur != forkHash {
		idx, ok, err := d.GetIndex(cur)
		if err != nil {
			return err
		}
		if !ok {
			return fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}

		undo, ok, err := d.GetUndo(cur)
		if err != nil {
			return err
		}
		if !ok || undo == nil {
			return fmt.Errorf("REORG_ERR_UNDO_MISSING")
		}

		// Apply undo atomically (DB batch), then advance manifest to parent as commit point.
		if err := d.db.Update(func(tx *bolt.Tx) error {
			bu := tx.Bucket(bucketUtxo)
			for _, c := range undo.Created {
				if err := bu.Delete(encodeOutpointKey(c)); err != nil {
					return err
				}
			}
			for _, s := range undo.Spent {
				val, err := encodeUtxoEntry(s.RestoredEntry)
				if err != nil {
					return err
				}
				if err := bu.Put(encodeOutpointKey(s.OutPoint), val); err != nil {
					return err
				}
			}
			return nil
		}); err != nil {
			return err
		}

		parentHash := idx.PrevHash
		parentIdx, ok, err := d.GetIndex(parentHash)
		if err != nil {
			return err
		}
		if !ok {
			return fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
		m := &Manifest{
			SchemaVersion: SchemaVersionV1,
			ChainIDHex:    d.manifest.ChainIDHex,

			TipHashHex:           hex32(parentHash),
			TipHeight:            parentIdx.Height,
			TipCumulativeWorkDec: parentIdx.CumulativeWork.Text(10),

			LastAppliedBlockHashHex: hex32(parentHash),
			LastAppliedHeight:       parentIdx.Height,
		}
		if err := d.SetManifest(m); err != nil {
			return err
		}
		cur = parentHash
	}

	// Connect fork+1 -> new tip.
	path, err := d.pathFromAncestor(forkHash, newTipHash)
	if err != nil {
		return err
	}
	for _, h := range path {
		blockBytes, ok, err := d.GetBlockBytes(h)
		if err != nil {
			return err
		}
		if !ok {
			return fmt.Errorf("REORG_ERR_BLOCK_MISSING")
		}
		// Direct connect: must extend current manifest tip by construction.
		decision, err := d.applyBlockAsNewTip(p, chainID, blockBytes, opts)
		if err != nil {
			// Mark INVALID_BODY on failure and stop reorg; applied tip stays consistent.
			idx, ok2, _ := d.GetIndex(h)
			if ok2 {
				idx.Status = BlockStatusInvalid
				_ = d.PutIndex(h, *idx)
			}
			return err
		}
		if decision != ApplyAppliedAsTip {
			return fmt.Errorf("reorg: unexpected decision %s", decision)
		}
	}
	return nil
}

func (d *DB) findForkPoint(oldTip [32]byte, newTip [32]byte) ([32]byte, error) {
	a := oldTip
	b := newTip

	ha, ok, err := d.GetIndex(a)
	if err != nil {
		return [32]byte{}, err
	}
	if !ok {
		return [32]byte{}, fmt.Errorf("REORG_ERR_INDEX_MISSING")
	}
	hb, ok, err := d.GetIndex(b)
	if err != nil {
		return [32]byte{}, err
	}
	if !ok {
		return [32]byte{}, fmt.Errorf("REORG_ERR_INDEX_MISSING")
	}

	for ha.Height > hb.Height {
		a = ha.PrevHash
		ha, ok, err = d.GetIndex(a)
		if err != nil {
			return [32]byte{}, err
		}
		if !ok {
			return [32]byte{}, fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
	}
	for hb.Height > ha.Height {
		b = hb.PrevHash
		hb, ok, err = d.GetIndex(b)
		if err != nil {
			return [32]byte{}, err
		}
		if !ok {
			return [32]byte{}, fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
	}
	for a != b {
		a = ha.PrevHash
		b = hb.PrevHash
		ha, ok, err = d.GetIndex(a)
		if err != nil {
			return [32]byte{}, err
		}
		if !ok {
			return [32]byte{}, fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
		hb, ok, err = d.GetIndex(b)
		if err != nil {
			return [32]byte{}, err
		}
		if !ok {
			return [32]byte{}, fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
	}
	return a, nil
}

// pathFromAncestor returns the hashes from ancestor's child up to tip (ascending height).
func (d *DB) pathFromAncestor(ancestor [32]byte, tip [32]byte) ([][32]byte, error) {
	if ancestor == tip {
		return nil, nil
	}
	// Walk back from tip to ancestor, then reverse.
	cur := tip
	out := make([][32]byte, 0, 16)
	for cur != ancestor {
		out = append(out, cur)
		idx, ok, err := d.GetIndex(cur)
		if err != nil {
			return nil, err
		}
		if !ok {
			return nil, fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
		cur = idx.PrevHash
		if cur == ([32]byte{}) {
			return nil, fmt.Errorf("REORG_ERR_INDEX_MISSING")
		}
	}
	// Reverse.
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}
	return out, nil
}

// applyBlockAsNewTip applies a block that is expected to directly extend the current applied tip.
func (d *DB) applyBlockAsNewTip(
	p crypto.CryptoProvider,
	chainID [32]byte,
	blockBytes []byte,
	opts ApplyOptions,
) (ApplyDecision, error) {
	// This is the "direct connect" half of ApplyBlockIfBestTip, without Stage0-3.
	block, err := consensus.ParseBlockBytes(blockBytes)
	if err != nil {
		return "", err
	}
	blockHash, err := consensus.BlockHeaderHash(p, block.Header)
	if err != nil {
		return "", err
	}
	tipHash, err := parseHex32(d.manifest.TipHashHex)
	if err != nil {
		return "", err
	}
	if block.Header.PrevBlockHash != tipHash {
		return "", fmt.Errorf("REORG_ERR_LINKAGE")
	}

	parentIndex, ok, err := d.GetIndex(tipHash)
	if err != nil {
		return "", err
	}
	if !ok {
		return "", fmt.Errorf("REORG_ERR_INDEX_MISSING")
	}
	height := parentIndex.Height + 1

	utxo, err := d.LoadUTXOSet()
	if err != nil {
		return "", err
	}
	ancestorHeaders, err := d.loadAncestorHeadersForParent(tipHash, height)
	if err != nil {
		return "", err
	}
	ctx := consensus.BlockValidationContext{
		Height:           height,
		AncestorHeaders:  ancestorHeaders,
		LocalTime:        opts.LocalTime,
		LocalTimeSet:     opts.LocalTimeSet,
		SuiteIDSLHActive: opts.SuiteIDSLHActive,
		HTLCV2Active:     opts.HTLCV2Active,
	}
	if err := consensus.ApplyBlock(p, chainID, &block, utxo, ctx); err != nil {
		return "", err
	}

	preUtxo, err := d.LoadUTXOSet()
	if err != nil {
		return "", err
	}
	undo, created, err := computeUndoForBlock(p, height, &block, preUtxo)
	if err != nil {
		return "", err
	}
	undo.Created = created

	createdEntries, err := computeCreatedEntries(p, height, &block)
	if err != nil {
		return "", err
	}
	undoBytes, err := encodeUndoRecord(undo)
	if err != nil {
		return "", err
	}

	idx, ok, err := d.GetIndex(blockHash)
	if err != nil {
		return "", err
	}
	if !ok {
		// If index wasn't set, compute minimal here.
		w, err := WorkFromTarget(block.Header.Target)
		if err != nil {
			return "", err
		}
		idx = &BlockIndexEntry{
			Height:         height,
			PrevHash:       tipHash,
			CumulativeWork: new(big.Int).Add(parentIndex.CumulativeWork, w),
			Status:         BlockStatusValid,
		}
	} else {
		idx.Height = height
		idx.PrevHash = tipHash
		idx.Status = BlockStatusValid
	}
	indexBytes, err := encodeIndexEntry(*idx)
	if err != nil {
		return "", err
	}

	if err := d.db.Update(func(tx *bolt.Tx) error {
		if err := tx.Bucket(bucketUndo).Put(blockHash[:], undoBytes); err != nil {
			return err
		}
		bu := tx.Bucket(bucketUtxo)
		for _, s := range undo.Spent {
			if err := bu.Delete(encodeOutpointKey(s.OutPoint)); err != nil {
				return err
			}
		}
		for _, ce := range createdEntries {
			val, err := encodeUtxoEntry(ce.Entry)
			if err != nil {
				return err
			}
			if err := bu.Put(encodeOutpointKey(ce.Point), val); err != nil {
				return err
			}
		}
		if err := tx.Bucket(bucketIndex).Put(blockHash[:], indexBytes); err != nil {
			return err
		}
		return nil
	}); err != nil {
		return "", err
	}

	m := &Manifest{
		SchemaVersion: SchemaVersionV1,
		ChainIDHex:    d.manifest.ChainIDHex,

		TipHashHex:           hex32(blockHash),
		TipHeight:            idx.Height,
		TipCumulativeWorkDec: idx.CumulativeWork.Text(10),

		LastAppliedBlockHashHex: hex32(blockHash),
		LastAppliedHeight:       idx.Height,
	}
	if err := d.SetManifest(m); err != nil {
		return "", err
	}
	return ApplyAppliedAsTip, nil
}

File: clients/rust/crates/rubin-store/src/reorg.rs
//! Reorg: disconnect/connect with undo logs.
//!
//! See `operational/RUBIN_REORG_DISCONNECT_CONNECT_v1.1.md`.

use std::path::Path;

use rubin_crypto::CryptoProvider;

use crate::db::Store;
use crate::keys::BlockIndexEntry;
use crate::manifest::{hex_encode, Manifest};
use crate::pipeline;

// ---------------------------------------------------------------------------
// Fork-point discovery (§4)
// ---------------------------------------------------------------------------

/// Find the common ancestor of `old_tip` and `new_tip` by walking back
/// via `prev_hash` links in the block index.
pub fn find_fork_point(
    store: &Store,
    old_tip: &[u8; 32],
    new_tip: &[u8; 32],
) -> Result<[u8; 32], String> {
    let mut a = *old_tip;
    let mut b = *new_tip;

    let get_index = |hash: &[u8; 32]| -> Result<BlockIndexEntry, String> {
        store
            .get_block_index(hash)?
            .ok_or_else(|| format!("REORG_ERR_INDEX_MISSING: {}", hex_encode(hash)))
    };

    let mut a_idx = get_index(&a)?;
    let mut b_idx = get_index(&b)?;

    // Equalize heights.
    while a_idx.height > b_idx.height {
        a = a_idx.prev_hash;
        a_idx = get_index(&a)?;
    }
    while b_idx.height > a_idx.height {
        b = b_idx.prev_hash;
        b_idx = get_index(&b)?;
    }

    // Walk both back until they meet.
    while a != b {
        a = a_idx.prev_hash;
        b = b_idx.prev_hash;
        a_idx = get_index(&a)?;
        b_idx = get_index(&b)?;
    }

    Ok(a)
}

// ---------------------------------------------------------------------------
// Disconnect (§7)
// ---------------------------------------------------------------------------

/// Disconnect one block: revert UTXO changes using its undo record.
/// Updates the manifest to the parent after successful revert.
fn disconnect_block(
    store: &Store,
    manifest: &mut Manifest,
    manifest_path: &Path,
    block_hash: &[u8; 32],
) -> Result<(), String> {
    let undo = store
        .get_undo(block_hash)?
        .ok_or_else(|| format!("REORG_ERR_UNDO_MISSING: {}", hex_encode(block_hash)))?;

    let idx = store
        .get_block_index(block_hash)?
        .ok_or_else(|| format!("REORG_ERR_INDEX_MISSING: {}", hex_encode(block_hash)))?;

    let wb = store.begin_write()?;

    // Delete created UTXOs.
    for outpoint in &undo.created {
        wb.delete_utxo(outpoint)?;
    }

    // Restore spent UTXOs.
    for item in &undo.spent {
        wb.put_utxo(&item.outpoint, &item.restored_entry)?;
    }

    wb.commit()?;

    // Compute parent's cumulative work.
    let parent_idx = store.get_block_index(&idx.prev_hash)?;
    let parent_work = parent_idx.map(|p| p.cumulative_work).unwrap_or(0);
    let parent_height = if idx.height > 0 { idx.height - 1 } else { 0 };

    // Update manifest to parent.
    manifest.update_tip(&hex_encode(&idx.prev_hash), parent_height, parent_work);
    manifest.save_atomic(manifest_path)?;

    Ok(())
}

// ---------------------------------------------------------------------------
// Execute reorg (§5)
// ---------------------------------------------------------------------------

/// Execute a full reorg from current tip to new_tip.
/// 1. Find fork point.
/// 2. Disconnect from old tip down to fork point.
/// 3. Connect from fork point up to new_tip.
pub fn execute_reorg(
    store: &Store,
    manifest: &mut Manifest,
    manifest_path: &Path,
    provider: &dyn CryptoProvider,
    chain_id: &[u8; 32],
    new_tip_hash: &[u8; 32],
) -> Result<(), String> {
    let old_tip = manifest.tip_hash_bytes()?;

    let fork_point = find_fork_point(store, &old_tip, new_tip_hash)?;

    // Collect blocks to disconnect (old_tip → fork_point, exclusive of fork).
    let mut disconnect_hashes = Vec::new();
    {
        let mut cursor = old_tip;
        while cursor != fork_point {
            disconnect_hashes.push(cursor);
            let idx = store
                .get_block_index(&cursor)?
                .ok_or_else(|| format!("REORG_ERR_INDEX_MISSING: {}", hex_encode(&cursor)))?;
            cursor = idx.prev_hash;
        }
    }

    // Collect blocks to connect (fork_point → new_tip, exclusive of fork).
    let mut connect_hashes = Vec::new();
    {
        let mut cursor = *new_tip_hash;
        while cursor != fork_point {
            connect_hashes.push(cursor);
            let idx = store
                .get_block_index(&cursor)?
                .ok_or_else(|| format!("REORG_ERR_INDEX_MISSING: {}", hex_encode(&cursor)))?;
            cursor = idx.prev_hash;
        }
        connect_hashes.reverse(); // ascending height order
    }

    // Disconnect in descending order (already in descending order from tip).
    for hash in &disconnect_hashes {
        disconnect_block(store, manifest, manifest_path, hash)?;
    }

    // Connect in ascending order.
    for hash in &connect_hashes {
        let block_bytes = store
            .get_block_bytes(hash)?
            .ok_or_else(|| format!("REORG_ERR_BLOCK_MISSING: {}", hex_encode(hash)))?;
        let block = rubin_consensus::parse_block_bytes(&block_bytes)
            .map_err(|e| format!("REORG_ERR_PARSE: {e}"))?;
        let block_hash_computed = provider.sha3_256(&rubin_consensus::block_header_bytes(&block.header))?;

        let idx = store
            .get_block_index(hash)?
            .ok_or_else(|| format!("REORG_ERR_INDEX_MISSING: {}", hex_encode(hash)))?;

        pipeline::connect_block(
            store,
            manifest,
            manifest_path,
            provider,
            chain_id,
            &block,
            &block_hash_computed,
            &block_bytes,
            idx.height,
            idx.cumulative_work,
        )?;
    }

    Ok(())
}
